ddl/index.go

```diff
@@ -700,6 +685,431 @@ func (w *worker) doBackfillIndexTaskInTxn(t table.Table, txn kv.Transaction, col
+	_, startHandle, err := tablecodec.DecodeRecordKey(keyRange.StartKey)
+	if err != nil {
+		return 0, 0, errors.Trace(err)
+	}
+	_, endHandle, err := tablecodec.DecodeRecordKey(keyRange.EndKey)
+	if err != nil {
+		return 0, 0, errors.Trace(err)
+	}
+
+	return startHandle, endHandle, nil
+}
+
+func closeAddIndexWorkers(workers []*addIndexWorker) {
+	for _, worker := range workers {
+		worker.close()
+	}
+}
+
+func (d *ddl) waitTaskResults(workers []*addIndexWorker, taskCnt int, totalAddedCount *int64, startHandle int64) (int64, int64, error) {
+	var (
+		addedCount int64
+		nextHandle = startHandle
+		firstErr   error
+	)
+	for i := 0; i < taskCnt; i++ {
+		worker := workers[i]
+		result := <-worker.resultCh
+		if firstErr == nil && result.err != nil {
+			firstErr = result.err
+			log.Infof("[ddl-reorg] error occured, notify remaining workers to return.")
+
+			// Notify remaining workers exit immediately.
+			// If we don't do this, we will need to wait the long run workers exit.
+			for j := i + 1; j < taskCnt; j++ {
+				workers[j].notifyNotRunnable()
+			}
+			// We should wait all working workers exits, any way.
+			continue
+		}
+
+		if result.err != nil {
+			log.Warnf("[ddl-reorg] worker[%v] return err:%v", i, result.err)
+		}
+
+		if firstErr == nil {
+			*totalAddedCount += int64(result.addedCount)
+			addedCount += int64(result.addedCount)
+			nextHandle = result.nextHandle
+
+			d.reorgCtx.setRowCountAndHandle(*totalAddedCount, nextHandle)
+		}
+	}
+
+	return nextHandle, addedCount, errors.Trace(firstErr)
+}
+
+// finishBatchTasks waits all the running worker return back result, there are taskCnt workers.
+func (d *ddl) finishBatchTasks(startTime time.Time, startHandle int64, reorgInfo *reorgInfo, totalAddedCount *int64, workers []*addIndexWorker, taskCnt int) error {
+	nextHandle, taskAddedCount, err := d.waitTaskResults(workers, taskCnt, totalAddedCount, startHandle)
+	elapsedTime := time.Since(startTime).Seconds()
+	if err == nil {
+		err = d.isReorgRunnable()
+	}
+
+	if err != nil {
+		// update the reorg handle that has been processed.
+		err1 := kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
+			return errors.Trace(reorgInfo.UpdateHandle(txn, nextHandle))
+		})
+		log.Warnf("[ddl-reorg] total added index for %d rows, this task [%d,%d) add index for %d failed %v, take time %v, update handle err %v",
+			*totalAddedCount, startHandle, nextHandle, taskAddedCount, err, elapsedTime, err1)
+		return errors.Trace(err)
+	}
+
+	// nextHandle will be updated periodically in runReorgJob, so no need to update it here.
+	d.reorgCtx.setRowCountAndHandle(*totalAddedCount, nextHandle)
+	metrics.BatchAddIdxHistogram.Observe(elapsedTime)
+	log.Infof("[ddl-reorg] total added index for %d rows, this task [%d,%d) added index for %d rows, take time %v",
+		*totalAddedCount, startHandle, nextHandle, taskAddedCount, elapsedTime)
+	return nil
+}
+
+func sendTaskToBackfillWorkers(workers []*addIndexWorker, batchTasks []*reorgIndexTask) {
```

Comment:
How about inline this function into `finishBatchTasks` and rename that to `backfillBatshTasks`?